File: DSA Practice/LeetCode/1114. Print in Order/solution.java
class Foo {
    private Lock lock = new ReentrantLock();
    private Condition secondCondition = lock.newCondition();
    private Condition thirdCondition = lock.newCondition();
    private boolean secondExecuted = false;
    private boolean thirdExecuted = false;

    public Foo() {
    }

    public void first(Runnable printFirst) throws InterruptedException {
        lock.lock();
        try {
            // printFirst.run() outputs "first". Do not change or remove this line.
            printFirst.run();
            secondExecuted = true;
            secondCondition.signal();
        } finally {
            lock.unlock();
        }
    }

    public void second(Runnable printSecond) throws InterruptedException {
        lock.lock();
        try {
            while (!secondExecuted) {
                secondCondition.await();
            }
            // printSecond.run() outputs "second". Do not change or remove this line.
            printSecond.run();
            thirdExecuted = true;
            thirdCondition.signal();
        } finally {
            lock.unlock();
        }
    }

    public void third(Runnable printThird) throws InterruptedException {
        lock.lock();
        try {
            while (!thirdExecuted) {
                thirdCondition.await();
            }
            // printThird.run() outputs "third". Do not change or remove this line.
            printThird.run();
        } finally {
            lock.unlock();
        }
    }
}

File: DSA Practice/LeetCode/1114. Print in Order/solution.py
import threading
class Foo:
    def __init__(self):
        self.f = threading.Lock()
        self.s = threading.Lock()
        self.s.acquire()
        self.t = threading.Lock()
        self.t.acquire()

    def first(self, printFirst: 'Callable[[], None]') -> None:
        with self.f:
            # printFirst() outputs "first". Do not change or remove this line.
            printFirst()
            self.s.release()

    def second(self, printSecond: 'Callable[[], None]') -> None:
        with self.s:
            # printSecond() outputs "second". Do not change or remove this line.
            printSecond()
            self.t.release()

    def third(self, printThird: 'Callable[[], None]') -> None:
        with self.t:
            # printThird() outputs "third". Do not change or remove this line.
            printThird()